docs/development_guidelines.md

```diff
@@ -0,0 +1,88 @@
+fix it.
+
+Before every push you ~~can~~ should run locally `./gradlew devCheck` that will both reformat the code and run the 
+unit tests.
+
+You can run the serverless locally with the command: `gradlew server:quarkusDev`
+
+### Windows
+
+Some hadoop related tests do not run on Windows (especially in the Windows CI) so you can find the disabled using junit5
+annotation `@DisabledOnOs(WINDOWS)`. If you want to provide a fix for them, feel free to open a PR.
+
+## Code formatting
+
+We use spotless to format and check the style of Java code. Spotless can be run through Gradle: 
+
+- to reformat: `./gradlew spotlessApply`
+- to check formatting: `./gradlew spotlessCheck`
+
+## Protocol
+
+ Whitefox protocol and original delta-sharing are kept under [`docs/protocol`](protocol). We use openapi 
+ specification to generate code for server and client. Furthermore, the protocol itself is validated using [spectral]
+ (https://stoplight.io/open-source/spectral), you can find spectral configuration at [`.spectral.yaml`](../.spectral.yaml).
+ 
+
+# Software Engineering guidelines
+
+Some golden rules that we use in this repo:
+
+- always prefer immutability, this helps writing correct concurrent code  
+- use constructor-based dependency injection when possible (i.e. do not annotate with `@Inject` any field) this 
+  makes testing beans easy without a CDI framework (that makes unit test much faster)
+- the "core" code should [make illegal states unrepresentable](https://khalilstemmler.com/articles/typescript-domain-driven-design/make-illegal-states-unrepresentable/)
+- never use `null`, use `Optional` to convey optionality, ignore IntelliJ "warning" when passing `Optional` to methods
+- throw exceptions and write beautiful error messages
+- use `@QuarkusTest` annotation only when you really need the server to be running, otherwise write a "simple" unit 
+  test
+- don't use `@SessionScoped` it will make our life miserable when trying to scale horizontally
+- [use `@ApplicationScoped` everywhere, `@Singleton` when `@ApplicationScoped` does not fit](https://quarkus.io/guides/cdi#applicationscoped-and-singleton-look-very-similar-which-one-should-i-choose-for-my-quarkus-application)
+- de-couple different layers of the application as much as possible:
+  - REST API classes should deal with auto-generated model classes, convert them (when needed) to the "core" 
+    representation and pass it to "services"
+  - services should not know anything about the outside world (i.e. REST API) and should deal with "core" models 
+    that do not depend on any given serialization format
+  - persistence is its own world too, if needed we can create "persistence-related" classes to use ORM tools but 
+    these should not interfere in any way with the core ones
+  - we will hand-craft (for now, who knows in the future) a lot of `mappers` that are function that go from one 
```

Comment:
Maybe use Mapstruct?

Reply:
I used it in the past, nice piece of software :) 
anyway so far it has been manageable to hand-write the mappers I would avoid yet another tool.